src/client/src/offset.rs

```diff
@@ -0,0 +1,273 @@
+/// a look:
+///
+/// ```text
+///                                        These events were added!
+///                                                               |
+///                                                      vvvvvvvvvv
+///         Partition Log: [ .., .., 22, 33, 44, 55, 66, 77, 88, 99 ]
+///       Absolute Offset:    0,  1,  2,  3,  4,  5,  6,  7,  8,  9
+///  FromBeginning Offset:            0,  1,  2,  3,  4,  5,  6,  7
+///        FromEnd Offset:    9,  8,  7,  6,  5,  4,  3,  2,  1,  0
+/// ```
+///
+/// # Example
+///
+/// All offsets must be constructed with a positive index. Negative
+/// numbers are meaningless for offsets and therefore are not allowed.
+/// Trying to construct an offset with a negative number will yield `None`.
+///
+/// ```
+/// use fluvio::Offset;
+/// let absolute_offset = Offset::absolute(5).unwrap();
+/// let offset_from_beginning = Offset::from_beginning(100).unwrap();
+/// let offset_from_end = Offset::from_end(10).unwrap();
+///
+/// // Negative offsets will give None
+/// assert_eq!(Offset::absolute(-10), None);
+/// assert_eq!(Offset::from_beginning(-15), None);
+/// assert_eq!(Offset::from_end(-20), None);
+/// ```
+#[derive(Debug, Clone, PartialEq)]
+pub struct Offset {
+    inner: OffsetInner,
+}
+
+impl Offset {
+    /// Creates an absolute offset with the given index
+    ///
+    /// The index must not be less than zero.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use fluvio::Offset;
+    /// assert!(Offset::absolute(100).is_some());
+    /// assert!(Offset::absolute(0).is_some());
+    /// assert!(Offset::absolute(-10).is_none());
+    /// ```
+    pub fn absolute(index: i64) -> Option<Offset> {
+        if index < 0 {
+            return None;
+        }
+        Some(Self {
+            inner: OffsetInner::Absolute(index),
+        })
+    }
+
+    /// Creates a relative offset starting at the beginning of the saved log
+    ///
+    /// A relative `FromBeginning` offset will not always match an `Absolute`
+    /// offset. In order to save space, Fluvio may sometimes delete events
+    /// from the beginning of the log. When this happens, the `FromBeginning`
+    /// relative offset starts counting from the first non-deleted log entry.
+    ///
+    /// ```text
+    ///                          These events were deleted!
+    ///                          |
+    ///                          vvvvvv
+    ///         Partition Log: [ .., .., 22, 33, 44, 55, 66 ]
+    ///       Absolute Offset:    0,  1,  2,  3,  4,  5,  6
+    ///  FromBeginning Offset:            0,  1,  2,  3,  4
+    /// ```
+    ///
+    /// The offset must not be less than zero.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use fluvio::Offset;
+    /// assert!(Offset::from_beginning(10).is_some());
+    /// assert!(Offset::from_beginning(0).is_some());
+    /// assert!(Offset::from_beginning(-10).is_none());
+    /// ```
+    pub fn from_beginning(offset: i64) -> Option<Offset> {
```

Comment:
why return option?

Reply:
This returns `Option` because if a user were to call `Offset::from_beginning(-10)`, that would be invalid. Having constructors return `Option`s is a way to create the invariant of "If it exists, then it's valid". This is to prevent bad offsets from being constructed.

Reply:
if you make parameter u64 then you don't need to worry about it.   I thought we are adding API so invalid argument can't be constructed

Reply:
If we made a function such as `from_beginning(offset: u64) -> Offset` (note the infallible return of an `Offset`), then that would be an inherently broken API, because not all `u64` values will produce a valid offset. If we are committed to representing this value internally as an `i64`, then we _must_ return an `Option<Offset>` in order to create a sound API. The choice now is whether to use an argument of `i64` and forbid negatives (this is the path I think we should take), or use an argument of `u64` and forbid values that overflow `i64::MAX`. I feel like the second choice feels much more arbitrary and will be more confusing for users.

Reply:
how about if we make offset u32?,  I don't believe there is use case we jump arbitrary offsets.  u32 should be big enough

Reply:
I think u32 will probably not be big enough. I think I had the same thought yesterday, I was like "a u32 should be half the size of u64, so it should have the same positive range as i64", but that's not correct. A u32 is several orders of magnitude smaller than even the positive range of an i64:

```
Max u64: 18446744073709551615
Max i64: 9223372036854775807
Max u32: 4294967295
Max i32: 2147483647
```

I think that we should just continue using `i64`

Reply:
Do you see any use case where someone wants to do relative offset bigger than u32?  Most likely, it will be small like last 100 or 1000.

Reply:
I can't immediately think of one, but I also don't think it's a good idea to prematurely restrict the things that users can do. If we change this to `from_beginning(offset: u32) -> Offset`, then there are potential valid cases that are no longer possible to express. I just feel like there's really not any harm in leaving it as `i64`

Reply:
if in that case, I don't see any advantage over previous API

Reply:
The previous API would not validate any of the values created in an `Offset` type. The previous definition was like this:
```rust
pub enum FetchOffset {
    Earliest(Option<i64>),
    Latest(Option<i64>),
    Offset(i64),
}
```

Since all variants of a public enum are also public, that means that I would be able to construct values such as these:
```rust
FetchOffset::Earliest(Some(-100))
FetchOffset::Latest(Some(-100))
FetchOffset::Offset(-100)
```

All of which are illegal values. This would place the responsibility of validation on every single internal piece of code that uses a `FetchOffset` value.

In contrast, the new `Offset` type can only be created via it's three constructors, `absolute`, `from_beginning`, and `from_end`. All of these functions return a `Result<Offet, FluvioError>` because there are values of `i64` which would be illegal for an offset (namely, all of the negative values of i64). The constructors will only yield an `Ok(Offset)` when the given value is valid (i.e. positive offset). Therefore, there is now an invariant on the `Offset` type which is that if an `Offset` value exists, it must be valid (i.e. non-negative). This is valuable because the internal implementation code may safely assume that all `Offset` values are valid and can eliminate repetitive checking and validating throughout the codebase. It also makes sure that a user is not given a false sense of security if they construct a `FetchOffset` that happens to be invalid but they don't find out about it until later on when they try to use it. The `Offset` API has more of a fail-fast mentality where the user will know immediately whether they created a correct value.